### PR TITLE
Fix incorrect handling of duration in v1 -> v2 arrow conversion

### PR DESCRIPTION
**Pull request by dispanser**
Our table slice conversion code for `arrow::v1` -> `arrow::v2` conversion did not specifically handle duration data. While the underlying phsysical representation is actually the same (`int64`), the types don't align and an assertion is triggered in the process.

We add a separate handler for duration that doesn't fail, and add a test based on some hand-crafted arrow-v1 table slice.

### :memo: Checklist

- [X] All user-facing changes have changelog entries.
- [X] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [X] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Simple enough, one commit.

### CODE REVIEW

File: libvast/src/arrow_table_slice.cpp [hunk 1]
```diff
@@ -1251,6 +1251,10 @@ upgrade_array_to_v2(const std::shared_ptr<arrow::Array>& arr, const type& t) {
       VAST_ASSERT(arr->type()->Equals(t.to_arrow_type()));
       return arr;
     },
+    [&](const duration_type&) -> std::shared_ptr<arrow::Array> {
+      // duration is backed by physical int64 array, but logical type differs
+      return arr;
```
Comment (dominiklohmann) [inline libvast/src/arrow_table_slice.cpp:1256]: You're deep in undefined behavior land here with pointer aliasing rules. I think the correct way to handle this is to create a new Array from the existing Array's underlying ArrayData.

```suggestion
      // Duration is backed by physical Int64 array, but logical type differs.
      auto result = arr->View(duration_type::to_arrow_type());
      VAST_ASSERT(result.ok(), result.status().ToString().c_str());
      return result.MoveValueUnsafe();
```

File: libvast/test/arrow_table_slice.cpp [hunk 2]
```diff
@@ -1050,6 +1050,18 @@ TEST(read_legacy_table_slice) {
   check_column(legacy_slice, 14, pattern_type{}, f3_pattern); // f11_2_2
 }
 
+TEST(convert_legacy_table_slice_all_types) {
+  auto bytes = unbox(vast::io::read(VAST_TEST_PATH "artifacts/table_slices/"
+                                                   "arrow-v1_all-types.bytes"));
+  auto legacy_slice
+    = table_slice{chunk::make(std::move(bytes)), table_slice::verify::yes};
+  fmt::print(stderr, "tp;was here {}, {}, {}\n", legacy_slice.columns(),
+             legacy_slice.rows(), bytes.size());
```
Comment (dominiklohmann) [inline libvast/test/arrow_table_slice.cpp:1059]: Leftover debug message.
```suggestion
```

File: libvast/test/arrow_table_slice.cpp [hunk 3]
```diff
@@ -1050,6 +1050,18 @@ TEST(read_legacy_table_slice) {
...
+  const auto& rb = to_record_batch(legacy_slice);
```
Comment (dominiklohmann) [inline libvast/test/arrow_table_slice.cpp:1060]: This record batch appears to be unused.

**Review by dominiklohmann — APPROVED** [review]
This is a really nice catch, I'm glad you found this _before_ we released the v2.0. I have some minor comments on the PR.

Thanks especially for the extensive test, IIUC that now covers all types.

File: libvast/test/arrow_table_slice.cpp (hunk 3 above)
Comment (dispanser) [reply libvast/test/arrow_table_slice.cpp:1058]: thanks, must be used to create a new table slice. added a commit.

**Pull request merged**